README.md
# eslint-config

A highly opinionated ESLint config.

## Getting started

Install from NPM, including peer dependencies, with `--save-dev`

```bash
npm i @nick-mazuk/eslint-config eslint --save-dev
```

Then, create a `.eslintrc` file in the root of your project with

```json
{
    "extends": ["@nick-mazuk/eslint-config"]
}
```

### Running the linter

Add this to your `package.json` file:

```json
"scripts": {
    "eslint": "eslint '**/*.{js,ts,tsx}' --quiet --fix",
    "check-types": "tsc --noemit",
    "lint": "npm run eslint && npm run check-types"
},
```

Then run

```bash
npm run lint
```

### Versions

This package has configs for different scenarios (more coming).

- **React**: use `@nick-mazuk/eslint-config/react`
- **NextJS**: use `@nick-mazuk/eslint-config/next`
- **Cypress**: use `@nick-mazuk/eslint-config/cypress`
- **Firebase functions**: use `@nick-mazuk/eslint-config/firebase-functions`

When using a more specific config, the base config is not needed. For instance, with NextJS, this is all that's required:

```json
{
    "extends": ["@nick-mazuk/eslint-config/next"]
}
```

## Integration with VS Code (recommended)

This is the easiest way to ensure ESLint is run every time you save a file, and that VS Code lints as you type, pointing out errors in real-time.

1. Create a new file `.vscode/settings.json`
2. Add this to the file:

```json
{
    "editor.codeActionsOnSave": {
        "source.fixAll.eslint": true
    },
    "eslint.run": "onType"
}
```

> **Note**: when you first open the repo with this ESLint configuration, it can take up to a minute for VS Code to start linting. You'll still be able to edit files in the meantime.

## Use with Prettier

This config has Prettier built-in. If you use Prettier with your code editor or use a `.prettierrc` file, there will likely be conflicts. It's best to remove any `.prettierrc` files you may have and disable Prettier in your code editor.

> **Rationale**: this ESLint config is highly opinionated and is a living config. As best practices evolve, this config will evolve. Since Prettier is 100% fixable, that means that your code will always follow this config and therefore always follow best practices. If you override the formatting with Prettier, you forgo any future updates in regard to Prettier formatting.

**Note**: If using the Svelte config, you need to setup Prettier manually. This is because the plugin `prettier-plugin-svelte` (which is needed to parse Svelte) [does not work with ESLint](https://github.com/sveltejs/prettier-plugin-svelte/issues/57).

## Use without TypeScript

Currently, all the configs require TypeScript. Perhaps in the future I'll create a TypeScript addon, but since TypeScript is just better than vanilla JS, vanilla JS is currently not supported.

## What rules are enabled?

In short, this config runs 500+ rules (and counting). Luckily, the source code is easy to read, so you should not have an issue reading the rules there.

## I found a bug, what do I do?

If you found a bug or rule conflicts, [submit an issue on GitHub](https://github.com/Nick-Mazuk/eslint-config/issues).


CHANGELOG.md
## [0.8.32](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.31...v0.8.32) (2022-04-01)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unicorn to v42 ([88eae8d](https://github.com/Nick-Mazuk/eslint-config/commit/88eae8dbc2b727dc7a1e73c243d54c3f5bf5e23e))

## [0.8.31](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.30...v0.8.31) (2022-03-29)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-sonarjs to ^0.13.0 ([b4d0519](https://github.com/Nick-Mazuk/eslint-config/commit/b4d0519bc4acbd47fd4b19bf82866969e6bb7d2f))

## [0.8.30](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.29...v0.8.30) (2022-02-18)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-sonarjs to ^0.12.0 ([c0d5959](https://github.com/Nick-Mazuk/eslint-config/commit/c0d595940acf398abd676f029c5ac288131573a3))

## [0.8.29](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.28...v0.8.29) (2022-02-16)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unicorn to v41 ([3794076](https://github.com/Nick-Mazuk/eslint-config/commit/37940760793890f33b836600994db80f2c702d90))

## [0.8.28](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.27...v0.8.28) (2022-01-24)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-jest to v26 ([3e84ae8](https://github.com/Nick-Mazuk/eslint-config/commit/3e84ae88704ea0fcf2a2ecda7a9053f306190aab))

## [0.8.27](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.26...v0.8.27) (2021-12-31)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unicorn to v40 ([8a1909d](https://github.com/Nick-Mazuk/eslint-config/commit/8a1909dbfe49989015546dee7f2a68f32ec15487))

## [0.8.26](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.25...v0.8.26) (2021-12-18)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-promise to v6 ([8a5dc96](https://github.com/Nick-Mazuk/eslint-config/commit/8a5dc96f6c5121104cd597a5cbf691289af3b5c2))

## [0.8.25](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.24...v0.8.25) (2021-12-01)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-sonarjs to ^0.11.0 ([3ee6e63](https://github.com/Nick-Mazuk/eslint-config/commit/3ee6e63f524c0336bf9f2bc04bcec806909a4dd7))

## [0.8.24](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.23...v0.8.24) (2021-11-21)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unicorn to v39 ([819a780](https://github.com/Nick-Mazuk/eslint-config/commit/819a78079580875cc8e2e897e4036f0705ec156b))

## [0.8.23](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.22...v0.8.23) (2021-11-10)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unused-imports to v2 ([000cf00](https://github.com/Nick-Mazuk/eslint-config/commit/000cf008cb1830ebc411c1879bdf23bf029188da))

## [0.8.22](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.21...v0.8.22) (2021-11-04)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unicorn to v38 ([57b8e81](https://github.com/Nick-Mazuk/eslint-config/commit/57b8e81e00c62ba2ad4ca7a7939693455008a14e))

## [0.8.21](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.20...v0.8.21) (2021-10-25)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-testing-library to v5 ([1c72266](https://github.com/Nick-Mazuk/eslint-config/commit/1c722668953f75143ce2b42a73e4f350721e5c52))

## [0.8.20](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.19...v0.8.20) (2021-10-11)


### Bug Fixes

* **deps:** bump: bump typescript-eslint monorepo to v5 ([6b9dc8f](https://github.com/Nick-Mazuk/eslint-config/commit/6b9dc8fb69b9f7e779c3fb61ae89252bc4800603))

## [0.8.19](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.18...v0.8.19) (2021-10-11)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unicorn to v37 ([03ab6b4](https://github.com/Nick-Mazuk/eslint-config/commit/03ab6b49eb783f9f1400f91adf1897532b29c35e))

## [0.8.18](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.17...v0.8.18) (2021-10-10)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-jest to v25 ([09efaa3](https://github.com/Nick-Mazuk/eslint-config/commit/09efaa3ff8080c1f8ef3e1fa5f50e0860b177822))

## [0.8.17](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.16...v0.8.17) (2021-09-11)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unicorn to v36 ([f2005ee](https://github.com/Nick-Mazuk/eslint-config/commit/f2005ee189cfefdd49b89a2c3a162a86255cc8e6))

## [0.8.16](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.15...v0.8.16) (2021-08-30)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-prettier to v4 ([3d64979](https://github.com/Nick-Mazuk/eslint-config/commit/3d6497973190c63e6fef1ad268c2f6d54182e605))

## [0.8.15](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.14...v0.8.15) (2021-08-13)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-sonarjs to ^0.10.0 ([3ac23ca](https://github.com/Nick-Mazuk/eslint-config/commit/3ac23ca277fbdeb5a92c9875d1dbffc18411a498))

## [0.8.14](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.13...v0.8.14) (2021-08-03)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unicorn to v35 ([bfb0d76](https://github.com/Nick-Mazuk/eslint-config/commit/bfb0d763110a00e84d9261fb1de7ca5ba4aac275))

## [0.8.13](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.12...v0.8.13) (2021-07-05)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-sonarjs to ^0.9.0 ([145ae4e](https://github.com/Nick-Mazuk/eslint-config/commit/145ae4eb0f2e31e2bb52898b0b4ac06103933289))

## [0.8.12](https://github.com/Nick-Mazuk/eslint-config/compare/v0.8.11...v0.8.12) (2021-06-28)


### Bug Fixes

* **deps:** bump: bump dependency eslint-plugin-unicorn to v34 ([fa1923a](https://github.com/Nick-Mazuk/eslint-config/commit/fa1923a6f6e01ed24d14b8e7935b65040612aa4e))


lib/base.js
module.exports = {
    env: {
        node: true,
    },
    plugins: [
        'import',
        'sonarjs',
        'no-secrets',
        'unicorn',
        'promise',
        'eslint-comments',
        'deprecation',
        'unused-imports',
    ],
    extends: [
        'eslint:recommended',
        'plugin:promise/recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:prettier/recommended',
    ],
    parser: '@typescript-eslint/parser',
    parserOptions: {
        sourceType: 'module',
        ecmaVersion: 2021,
        project: './tsconfig.json',
    },
    ignorePatterns: ['coverage/*'],
    rules: {
        'prettier/prettier': [
            'error',
            {
                // also defined in svelte config -- change values there if updated
                singleQuote: true,
                printWidth: 100,
                tabWidth: 4,
                useTabs: false,
                semi: false,
                quoteProps: 'as-needed',
                jsxSingleQuote: true,
                trailingComma: 'es5',
                bracketSpacing: true,
                jsxBracketSameLine: false,
                arrowParens: 'always',
                proseWrap: 'never',
                htmlWhitespaceSensitivity: 'css',
                endOfLine: 'lf',
                embeddedLanguageFormatting: 'auto',
            },
        ],
        'linebreak-style': ['error', 'unix'],
        'no-await-in-loop': 'error',
        'no-console': 'error',
        'no-extra-parens': 'off',
        'no-loss-of-precision': 'error',
        'no-promise-executor-return': 'error',
        'no-template-curly-in-string': 'error',
        'no-unreachable-loop': 'error',
        'no-useless-backreference': 'error',
        'require-atomic-updates': 'error',
        'accessor-pairs': 'error',
        'array-callback-return': 'error',
        'class-methods-use-this': 'error',
        'consistent-return': 'error',
        'default-case': 'error',
        'default-case-last': 'error',
        'default-param-last': 'error',
        curly: ['error', 'multi-or-nest', 'consistent'],
        'dot-location': ['error', 'property'],
        'dot-notation': 'error',
        eqeqeq: 'error',
        'grouped-accessor-pairs': ['error', 'getBeforeSet'],
        'guard-for-in': 'warn',
        'max-classes-per-file': ['error', 1],
        'no-caller': 'error',
        'no-alert': 'error',
        'no-constructor-return': 'error',
        'no-div-regex': 'error',
        'no-else-return': 'error',
        'no-eval': 'error',
        'no-extend-native': 'error',
        'no-extra-bind': 'error',
        'no-extra-label': 'error',
        'no-floating-decimal': 'error',
        'no-implicit-coercion': 'error',
        'no-implicit-globals': 'error',
        'no-implied-eval': 'error',
        'no-invalid-this': 'error',
        'no-iterator': 'error',
        'no-labels': 'error',
        'no-lone-blocks': 'error',
        'no-loop-func': 'error',
        'no-magic-numbers': [
            'warn',
            { ignore: [-1, 0, 1, 2, 100, 200, 204, 300, 404, 500, 1000, 1024] },
        ],
        'no-multi-str': 'error',
        'no-new': 'error',
        'no-new-func': 'error',
        'no-new-wrappers': 'error',
        'no-octal-escape': 'error',
        'no-param-reassign': 'error',
        'no-proto': 'error',
        'no-return-assign': 'error',
        'no-return-await': 'error',
        'no-script-url': 'error',
        'no-self-compare': 'error',
        'no-sequences': 'error',
        'no-throw-literal': 'error',
        'no-unmodified-loop-condition': 'error',
        'no-unused-expressions': ['error', { allowTernary: true }],
        'no-useless-call': 'error',
        'no-useless-concat': 'error',
        'no-useless-return': 'error',
        'no-void': 'error',
        'no-warning-comments': 'warn',
        'prefer-promise-reject-errors': 'error',
        'prefer-regex-literals': 'error',
        'require-await': 'error',
        'require-unicode-regexp': 'error',
        'wrap-iife': 'error',
        yoda: 'error',
        strict: ['error', 'never'],
        'init-declarations': ['error', 'always'],
        'no-label-var': 'error',
        'no-shadow': 'error',
        'no-undef-init': 'error',
        'no-undefined': 'error',
        'no-use-before-define': 'error',
        camelcase: 'error',
        'capitalized-comments': ['error', 'never'],
        'func-name-matching': 'error',
        'func-style': 'error',
        'id-length': ['error', { max: 24, min: 1, properties: 'never' }],
        'line-comment-position': 'error',
        'lines-around-comment': [
            'error',
            {
                beforeBlockComment: true,
                afterBlockComment: true,
                beforeLineComment: true,
                afterLineComment: false,
            },
        ],
        'lines-between-class-members': 'error',
        'max-lines-per-function': 'error',
        'multiline-comment-style': ['error', 'bare-block'],
        'new-cap': 'error',
        'new-parens': 'error',
        'no-array-constructor': 'error',
        'no-bitwise': 'error',
        'no-continue': 'error',
        'no-inline-comments': 'error',
        'no-lonely-if': 'error',
        'no-multi-assign': 'error',
        'no-negated-condition': 'error',
        'no-nested-ternary': 'error',
        'no-new-object': 'error',
        'no-unneeded-ternary': 'error',
        'no-restricted-imports': [
            'error',
            {
                paths: ['moment', 'lodash'],
            },
        ],
        'one-var': ['error', 'never'],
        'operator-assignment': 'error',
        'prefer-exponentiation-operator': 'error',
        'prefer-object-spread': 'error',
        'spaced-comment': 'error',
        'wrap-regex': 'error',
        'no-confusing-arrow': 'error',
        'no-useless-computed-key': 'error',
        'no-useless-constructor': 'error',
        'no-useless-rename': 'error',
        'no-var': 'error',
        'prefer-const': 'error',
        'prefer-destructuring': [
            'error',
            {
                AssignmentExpression: {
                    array: false,
                    object: false,
                },
            },
        ],
        'prefer-spread': 'error',
        'prefer-template': 'error',
        'symbol-description': 'error',
        'template-curly-spacing': 'error',
        'import/order': [
            'error',
            {
                groups: ['builtin', 'external', 'internal'],
                pathGroups: [
                    { pattern: 'react', group: 'external', position: 'before' },
                    { pattern: 'component/**', group: 'internal', position: 'after' },
                    { pattern: 'lib/**', group: 'internal', position: 'after' },
                    { pattern: 'classnames', group: 'external', position: 'before' },
                    { pattern: 'next/**', group: 'external', position: 'before' },
                ],
                pathGroupsExcludedImportTypes: ['react'],
                'newlines-between': 'always',
                alphabetize: {
                    order: 'asc',
                    caseInsensitive: true,
                },
            },
        ],
        'import/newline-after-import': ['error', { count: 1 }],
        'import/first': 'error',
        'import/exports-last': 'warn',
        'import/no-absolute-path': 'error',
        'import/no-default-export': 'error',
        'import/no-duplicates': 'error',
        'import/no-extraneous-dependencies': 'warn',
        'import/no-mutable-exports': 'error',
        'import/no-named-as-default-member': 'warn',
        'import/no-named-as-default': 'error',
        'import/no-relative-parent-imports': 'warn',
        'import/no-self-import': 'error',
        'import/no-commonjs': 'error',
        'import/no-cycle': ['error', { ignoreExternal: true }],
        'import/no-useless-path-segments': 'error',
        'import/no-webpack-loader-syntax': 'error',
        'import/unambiguous': 'error',
        'sonarjs/no-all-duplicated-branches': 'error',
        'sonarjs/no-element-overwrite': 'warn',
        'sonarjs/no-identical-conditions': 'error',
        'sonarjs/no-identical-expressions': 'error',
        'sonarjs/cognitive-complexity': ['error', 10],
        'sonarjs/max-switch-cases': 'error',
        'sonarjs/no-collapsible-if': 'error',
        'sonarjs/no-collection-size-mischeck': 'error',
        'sonarjs/no-duplicated-branches': 'error',
        'sonarjs/no-identical-functions': 'error',
        'sonarjs/no-inverted-boolean-check': 'error',
        'sonarjs/no-redundant-boolean': 'error',
        'sonarjs/no-redundant-jump': 'error',
        'sonarjs/no-small-switch': 'error',
        'sonarjs/no-unused-collection': 'error',
        'sonarjs/no-useless-catch': 'error',
        'sonarjs/prefer-immediate-return': 'error',
        'sonarjs/prefer-object-literal': 'error',
        'sonarjs/prefer-single-boolean-return': 'error',
        'sonarjs/prefer-while': 'error',
        'no-secrets/no-secrets': [
            'error',
            {
                ignoreContent: ['^https://www.youtube.com/', '^https://fonts.googleapis.com/'],
                // from https://github.com/dxa4481/truffleHogRegexes/blob/master/truffleHogRegexes/regexes.json
                additionalRegexes: {
                    'Slack Token': '(xox[p|b|o|a]-[0-9]{12}-[0-9]{12}-[0-9]{12}-[a-z0-9]{32})',
                    'RSA private key': '-----BEGIN RSA PRIVATE KEY-----',
                    'SSH (DSA) private key': '-----BEGIN DSA PRIVATE KEY-----',
                    'SSH (EC) private key': '-----BEGIN EC PRIVATE KEY-----',
                    'PGP private key block': '-----BEGIN PGP PRIVATE KEY BLOCK-----',
                    'Amazon AWS Access Key ID': 'AKIA[0-9A-Z]{16}',
                    'Amazon MWS Auth Token':
                        'amzn\\.mws\\.[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}',
                    'AWS API Key': 'AKIA[0-9A-Z]{16}',
                    'Facebook Access Token': 'EAACEdEose0cBA[0-9A-Za-z]+',
                    'Facebook OAuth':
                        '[f|F][a|A][c|C][e|E][b|B][o|O][o|O][k|K].*[\'|"][0-9a-f]{32}[\'|"]',
                    GitHub: '[g|G][i|I][t|T][h|H][u|U][b|B].*[\'|"][0-9a-zA-Z]{35,40}[\'|"]',
                    'Generic API Key':
                        '[a|A][p|P][i|I][_]?[k|K][e|E][y|Y].*[\'|"][0-9a-zA-Z]{32,45}[\'|"]',
                    'Generic Secret':
                        '[s|S][e|E][c|C][r|R][e|E][t|T].*[\'|"][0-9a-zA-Z]{32,45}[\'|"]',
                    'Google API Key': 'AIza[0-9A-Za-z\\-_]{35}',
                    'Google Cloud Platform API Key': 'AIza[0-9A-Za-z\\-_]{35}',
                    'Google Cloud Platform OAuth':
                        '[0-9]+-[0-9A-Za-z_]{32}\\.apps\\.googleusercontent\\.com',
                    'Google Drive API Key': 'AIza[0-9A-Za-z\\-_]{35}',
                    'Google Drive OAuth':
                        '[0-9]+-[0-9A-Za-z_]{32}\\.apps\\.googleusercontent\\.com',
                    'Google (GCP) Service-account': '"type": "service_account"',
                    'Google Gmail API Key': 'AIza[0-9A-Za-z\\-_]{35}',
                    'Google Gmail OAuth':
                        '[0-9]+-[0-9A-Za-z_]{32}\\.apps\\.googleusercontent\\.com',
                    'Google OAuth Access Token': 'ya29\\.[0-9A-Za-z\\-_]+',
                    'Google YouTube API Key': 'AIza[0-9A-Za-z\\-_]{35}',
                    'Google YouTube OAuth':
                        '[0-9]+-[0-9A-Za-z_]{32}\\.apps\\.googleusercontent\\.com',
                    'Heroku API Key':
                        '[h|H][e|E][r|R][o|O][k|K][u|U].*[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}',
                    'MailChimp API Key': '[0-9a-f]{32}-us[0-9]{1,2}',
                    'Mailgun API Key': 'key-[0-9a-zA-Z]{32}',
                    'Password in URL':
                        '[a-zA-Z]{3,10}://[^/\\s:@]{3,20}:[^/\\s:@]{3,20}@.{1,100}["\'\\s]',
                    'PayPal Braintree Access Token':
                        'access_token\\$production\\$[0-9a-z]{16}\\$[0-9a-f]{32}',
                    'Picatic API Key': 'sk_live_[0-9a-z]{32}',
                    'Slack Webhook':
                        'https://hooks.slack.com/services/T[a-zA-Z0-9_]{8}/B[a-zA-Z0-9_]{8}/[a-zA-Z0-9_]{24}',
                    'Stripe API Key': 'sk_live_[0-9a-zA-Z]{24}',
                    'Stripe Restricted API Key': 'rk_live_[0-9a-zA-Z]{24}',
                    'Square Access Token': 'sq0atp-[0-9A-Za-z\\-_]{22}',
                    'Square OAuth Secret': 'sq0csp-[0-9A-Za-z\\-_]{43}',
                    'Twilio API Key': 'SK[0-9a-fA-F]{32}',
                    'Twitter Access Token':
                        '[t|T][w|W][i|I][t|T][t|T][e|E][r|R].*[1-9][0-9]+-[0-9a-zA-Z]{40}',
                    'Twitter OAuth':
                        '[t|T][w|W][i|I][t|T][t|T][e|E][r|R].*[\'|"][0-9a-zA-Z]{35,44}[\'|"]',
                },
            },
        ],
        'unicorn/better-regex': 'error',
        'unicorn/catch-error-name': 'error',
        'unicorn/consistent-function-scoping': 'error',
        'unicorn/error-message': 'error',
        'unicorn/escape-case': 'error',
        'unicorn/explicit-length-check': 'error',
        'unicorn/filename-case': [
            'error',
            {
                case: 'kebabCase',
            },
        ],
        'unicorn/no-abusive-eslint-disable': 'error',
        'unicorn/no-array-instanceof': 'error',
        'unicorn/no-fn-reference-in-iterator': 'error',
        'unicorn/no-for-loop': 'error',
        'unicorn/no-hex-escape': 'error',
        'unicorn/no-new-buffer': 'error',
        'unicorn/no-null': 'warn',
        'unicorn/no-object-as-default-parameter': 'error',
        'unicorn/no-process-exit': 'error',
        'unicorn/no-reduce': 'error',
        'unicorn/no-unreadable-array-destructuring': 'error',
        'unicorn/no-unsafe-regex': 'error',
        'unicorn/no-useless-undefined': 'error',
        'unicorn/no-zero-fractions': 'error',
        'unicorn/prefer-add-event-listener': 'error',
        'unicorn/prefer-array-find': 'error',
        'unicorn/prefer-dataset': 'error',
        'unicorn/prefer-event-key': 'error',
        'unicorn/prefer-flat-map': 'error',
        'unicorn/prefer-includes': 'error',
        'unicorn/prefer-modern-dom-apis': 'error',
        'unicorn/prefer-negative-index': 'error',
        'unicorn/prefer-node-append': 'error',
        'unicorn/prefer-node-remove': 'error',
        'unicorn/prefer-optional-catch-binding': 'error',
        'unicorn/prefer-query-selector': 'error',
        'unicorn/prefer-replace-all': 'error',
        'unicorn/prefer-set-has': 'error',
        'unicorn/prefer-spread': 'error',
        'unicorn/prefer-starts-ends-with': 'error',
        'unicorn/prefer-string-slice': 'error',
        'unicorn/prefer-text-content': 'error',
        'unicorn/prefer-trim-start-end': 'error',
        'unicorn/prefer-type-error': 'error',
        'unicorn/prevent-abbreviations': [
            'error',
            {
                replacements: {
                    props: {
                        properties: false,
                    },
                    prop: {
                        property: false,
                    },
                    temp: false,
                    res: false,
                    req: false,
                    ref: false,
                    dev: false,
                    prod: false,
                    docs: false,
                    args: {
                        arguments: false,
                    },
                },
            },
        ],
        'unicorn/throw-new-error': 'error',
        'eslint-comments/disable-enable-pair': ['error', { allowWholeFile: true }],
        'eslint-comments/no-duplicate-disable': 'error',
        'eslint-comments/no-unlimited-disable': 'error',
        'eslint-comments/no-unused-disable': 'error',
        'eslint-comments/no-unused-enable': 'error',
        // used to ensure specific rules cannot be disabled
        'eslint-comments/no-restricted-disable': ['error'],
        'eslint-comments/require-description': 'error',
        'deprecation/deprecation': 'warn',
        '@typescript-eslint/array-type': 'error',
        '@typescript-eslint/consistent-type-assertions': 'error',
        '@typescript-eslint/consistent-type-definitions': ['error', 'type'],
        '@typescript-eslint/consistent-type-imports': 'error',
        '@typescript-eslint/method-signature-style': ['error', 'property'],
        '@typescript-eslint/no-confusing-non-null-assertion': 'error',
        '@typescript-eslint/no-dynamic-delete': 'error',
        '@typescript-eslint/no-floating-promises': 'error',
        '@typescript-eslint/no-invalid-void-type': 'error',
        '@typescript-eslint/no-implicit-any-catch': 'error',
        '@typescript-eslint/no-namespace': [
            'error',
            { allowDeclarations: true, allowDefinitionFiles: true },
        ],
        '@typescript-eslint/no-unnecessary-boolean-literal-compare': 'error',
        '@typescript-eslint/no-unnecessary-condition': 'error',
        '@typescript-eslint/no-unnecessary-qualifier': 'error',
        '@typescript-eslint/no-unnecessary-type-arguments': 'error',
        '@typescript-eslint/prefer-for-of': 'error',
        '@typescript-eslint/prefer-includes': 'error',
        '@typescript-eslint/prefer-nullish-coalescing': 'error',
        '@typescript-eslint/prefer-optional-chain': 'error',
        '@typescript-eslint/prefer-reduce-type-parameter': 'error',
        '@typescript-eslint/prefer-string-starts-ends-with': 'error',
        '@typescript-eslint/prefer-ts-expect-error': 'error',
        'unused-imports/no-unused-imports-ts': 'error',
    },
    overrides: [
        {
            files: ['**/*.d.ts'],
            rules: {
                'import/unambiguous': 'off',
            },
        },
    ],
}
